Add surface angular flux I/O support

rodriguez-jay opened the pull request:
- Added functionality to export surface angular fluxes at a specified boundary id or user provided surface id. 
- Added functionality for computing the response function along a surface provided forward and adjoint surfaces.

quocdang1998 commented on python/lib/solver.cc:
```diff
@@ -403,6 +403,77 @@ WrapLBS(py::module& slv)
+    {
+      // Map boundary names
+      std::map<std::string, uint64_t> allowed_bd_names = LBSProblem::supported_boundary_names;
+      std::map<std::uint64_t, std::string> allowed_bd_ids = LBSProblem::supported_boundary_ids;
```
replace this with `const std::map<...>&`. You don't need to copy these maps to use them.

andrsd commented on modules/linear_boltzmann_solvers/lbs_problem/io/angular_flux_io.cc:
```diff
@@ -228,9 +228,18 @@ LBSSolverIO::ReadAngularFluxes(
     H5ReadDataset1D<double>(file_id, group_name + "/values", values);
     for (uint64_t c = 0; c < file_num_local_cells; ++c)
     {
+      // bool isBndry = false; 
```
remove if this is not needed

andrsd commented on modules/linear_boltzmann_solvers/lbs_problem/io/angular_flux_io.cc:
```diff
@@ -239,9 +248,331 @@ LBSSolverIO::ReadAngularFluxes(
+
+  log.Log() << "Writing surface angular flux to " << file_base;
+
+  std::vector<std::vector<double>>& psi = do_problem.GetPsiNewLocal();
```
```
auto & psi = ...
```

rodriguez-jay committed:
Add surface angular flux response support

rodriguez-jay commented:
- Added surface angular-flux write/read support for boundary and internal surfaces.
- Exposed the new surface angular-flux functionality through the Python bindings.
- Cleaned up the implementation and resolved the clang-tidy.
- Added response_2d_surf.py as an adjoint regression test, including checks for:
  - Forward and adjoint QoIs
  - Forward surface angular flux
  - Adjoint surface importance

wdhawkins commented on modules/linear_boltzmann_solvers/discrete_ordinates_problem/io/angular_flux_io.cc:
```diff
@@ -242,5 +245,427 @@ DiscreteOrdinatesProblemIO::ReadAngularFluxes(
+      not H5CreateAttribute(file.Id(), group_name + "/num_groups", num_gs_groups),
+      "Failed to write the group count for " + group_name + ".");
+
+    for (const auto& surf_id : surface_tags)
```
This loop only writes h5 groups for surface tags that exist on the current rank. However, `ReadSurfaceAngularFluxes` tries to read every requested surface tag from every h5 file. If a surface exists only on some ranks, ranks without local faces for that surface will not have the corresponding data in their h5 files, causing the read to fail. We should either write empty datasets for all surfaces on every rank, or the reading routine shouldn't fail on missing surface data.

It also looks like we only validate groupset, group, and direction counts. We should also compare cell ids, face node counts, etc. I believe `ReadAngularFluxes` does pretty extensive checks. You might look at it to see what it does.

wdhawkins commented on modules/linear_boltzmann_solvers/discrete_ordinates_problem/io/angular_flux_io.cc:
```diff
@@ -242,5 +245,427 @@ DiscreteOrdinatesProblemIO::ReadAngularFluxes(
+                                          : axis == "y" ? Vector3{0.0, 1.0, 0.0}
+                                                        : Vector3{0.0, 0.0, 1.0};
+              const auto alignment = face.normal.Dot(global_norm);
+              surf_name = surf_id + (alignment > 0 ? "_u" : "_d");
```
What happens if a face matches multiple definitions? Here it looks like we would just overwrite each match with the next.

wdhawkins commented on modules/linear_boltzmann_solvers/discrete_ordinates_problem/io/angular_flux_io.cc:
```diff
@@ -242,5 +245,427 @@ DiscreteOrdinatesProblemIO::ReadAngularFluxes(
+          const auto id = std::find(bndry_ids.begin(), bndry_ids.end(), face.neighbor_id);
+          if (id != bndry_ids.end())
+          {
+            surf_name = allowed_bd_ids.at(*id);
```
See comment at line 370.

quocdang1998 commented on python/lib/solver.cc:
```diff
@@ -403,6 +403,77 @@ WrapLBS(py::module& slv)
+    Parameters
+    ----------
+    file_base: str
+        File basename.
```
If I understand correctly, you expect `surfaces` to be either a `List[str]` (`boundary_surfaces`) or a `List[Tuple[str,Dict]]` (`internal_surfaces`).

However, from the C++ API, `WriteSurfaceAngularFluxes` can accept both form, which mean that on the C++ API, user is allowed to provide both a list of surface names and a list of `(name, {axis: value})`. This feature is not available on the Python interface. Is this intentional or just a bug?

Regarding the second case (`internal_surfaces`), change the type of this argument in C++ API to `std::map<std::string, std::pair<std::string, double>>`, which is much more readable than `std::vector<std::pair<...>>`. In the Python interface, allow this argument to accept either a `List[str]` or a `Dict[str,Tuple[str,float]]`. At the end, a vector of pairs is simply a map.
